### PR TITLE
Restore unsupported tools-version error

### PR DESCRIPTION
**Pull request by neonichu**
We lost this in the 5.4 refactoring of tools-version parsing.

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
BTW, potentially interesting detail I noticed: we allow a missing comment to mean "this is version 3" but that's actually not a complete representation of what version 3 allowed, because e.g. one could have a random comment at the top of the manifest. Since we haven't heard about this so far, it's probably fine, though 🤣

**Comment by abertelrud** [comment]
Hm, that's unfortunate.  Is there a unit test that would catch this?

**Review by abertelrud — APPROVED** [review]

**Comment by abertelrud** [comment]
> BTW, potentially interesting detail I noticed: we allow a missing comment to mean "this is version 3" but that's actually not a complete representation of what version 3 allowed, because e.g. one could have a random comment at the top of the manifest. Since we haven't heard about this so far, it's probably fine, though 🤣

Could people be suffering in silence?  Would it break other packages to restore this?  Not necessarily immediately, but of course.  Or maybe it's better just to move on — do we have a sense of whether there are any v3 packages still in use?

**Comment by neonichu** [comment]
> Could people be suffering in silence? Would it break other packages to restore this? Not necessarily immediately, but of course. Or maybe it's better just to move on — do we have a sense of whether there are any v3 packages still in use?

To clarify: v3 hasn't been supported since Swift 5.0 as far as I can see, but what is supported is having a v3 manifest still present (as version-specific or main manifest) while there's also a newer manifest which then gets picked. If the regular manifest is a v3 one, we would only assume it's v3 if it has no comment at the top of the file at all which is technically inaccurate -- in other cases, we would error out.

File: Sources/PackageLoading/ToolsVersionParser.swift [hunk 1]
```diff
@@ -85,7 +85,17 @@ public struct ToolsVersionParser {
             throw Error.nonUTF8EncodedManifest(path: manifestPath)
         }
 
-        return try self.parse(utf8String: manifestContentsDecodedWithUTF8)
+        do {
+            return try self.parse(utf8String: manifestContentsDecodedWithUTF8)
+        }
+        catch {
```
Comment (tomerd) [inline Sources/PackageLoading/ToolsVersionParser.swift:91]: catch specific error instead of nested if

Comment (neonichu) [reply Sources/PackageLoading/ToolsVersionParser.swift:91]: Does that work when we need to catch on specific arguments? I thought that was only possible at the type-level.

**Comment by abertelrud** [comment]
> To clarify: v3 hasn't been supported since Swift 5.0 as far as I can see, but what is supported is having a v3 manifest still present (as version-specific or main manifest) while there's also a newer manifest which then gets picked. If the regular manifest is a v3 one, we would only assume it's v3 if it has no comment at the top of the file at all which is technically inaccurate -- in other cases, we would error out.

Oh I see.  I misunderstood.

File: Sources/PackageLoading/ToolsVersionParser.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageLoading/ToolsVersionParser.swift:91]: this should work:

```swift
do {
  return try self.parse(utf8String: manifestContentsDecodedWithUTF8)
} catch Error.malformedToolsVersionSpecification(.commentMarker(.isMissing)) {
  throw UnsupportedToolsVersion(packageIdentity: .init(path: manifestPath), currentToolsVersion: .current, packageToolsVersion: .v3)
}
```

**Review by tomerd — APPROVED** [review]

File: Sources/PackageLoading/ToolsVersionParser.swift [hunk 2]
```diff
@@ -654,7 +658,7 @@ extension ManifestLoader {
                 regularManifestToolsVersion = try ToolsVersionParser.parse(manifestPath: regularManifest, fileSystem: fileSystem)
             }
             catch {
-                if case ToolsVersionParser.Error.malformedToolsVersionSpecification(.commentMarker(.isMissing)) = error {
+                if let unsupportedError = error as? UnsupportedToolsVersion, unsupportedError.packageToolsVersion == .v3 {
```
Comment (tomerd) [inline Sources/PackageLoading/ToolsVersionParser.swift:661]: heh, I think we can do the same catch style here too

Comment (neonichu) [reply Sources/PackageLoading/ToolsVersionParser.swift:661]: Good point

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Pull request merged**